### PR TITLE
[ScriptSystem] Allow ScriptSystem to be derived from.

### PR DESCRIPTION
**Pull request by Eideren**
Minor changes to allow for game-specific implementation of the 'ScriptSystem'. 
This enables users to implement custom ScriptComponent events, execution culling, etc. While requiring little to no maintenance on our part.

### CODE REVIEW

File: sources/engine/Xenko.Engine/Engine/Processors/ScriptSystem.cs [hunk 1]
```diff
@@ -127,7 +128,7 @@ public override void Update(GameTime gameTime)
         /// Allows to wait for next frame.
         /// </summary>
         /// <returns>ChannelMicroThreadAwaiter&lt;System.Int32&gt;.</returns>
-        public ChannelMicroThreadAwaiter<int> NextFrame()
+        public virtual ChannelMicroThreadAwaiter<int> NextFrame()
```
Comment (Kryptos-FR) [inline sources/engine/Xenko.Engine/Engine/Processors/ScriptSystem.cs]: Virtual methods are more expensive than non-virtual and can prevent inlining.

While making `Add` and `Remove` virtual will probably have a low impact, `NextFrame()` is called very often, hence its impact is non-trivial.

**Comment by Eideren** [comment]
Okay.

**Comment by xen2** [comment]
Few remarks:
* Since `NextFrame()` is something called from user script, it shouldn't be hard for you to create a custom type that provide all the sync points you want? (it could even use `NextFrame()` under the hood)
* You mention you want to do custom events, so it's not only about overriding `NextFrame()` but also providing new functions? (which can also be done in an outside class)
I mostly want to make sure you can do everything you want in an external class or if the unseal/virtual would make some new scenario possible.
* This code will be replaced/obsolete soon (#20). `NextFrame` was OK to start with but we need much more control.
Interested in your use case, should take them into account during #20 redesign.

**Comment by Kryptos-FR** [comment]
> Kryptos-FR provided enough good reasons for me to drop this one off

@Eideren I am not always right and I could be missing some key points.. Don't be so prompt to delete your work until the discussion has reached a conclusion :wink:

**Pull request closed without merging**